src/main/java/sx/blah/discord/handle/impl/obj/Category.java

```diff
@@ -0,0 +1,266 @@
+import sx.blah.discord.api.internal.DiscordClientImpl;
+import sx.blah.discord.api.internal.DiscordEndpoints;
+import sx.blah.discord.api.internal.DiscordUtils;
+import sx.blah.discord.api.internal.json.objects.OverwriteObject;
+import sx.blah.discord.api.internal.json.requests.ChannelEditRequest;
+import sx.blah.discord.handle.obj.*;
+import sx.blah.discord.util.LogMarkers;
+import sx.blah.discord.util.PermissionUtils;
+import sx.blah.discord.util.cache.Cache;
+import sx.blah.discord.util.cache.LongMap;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.EnumSet;
+import java.util.List;
+import java.util.stream.Collectors;
+
+public class Category implements ICategory {
+
+	private final DiscordClientImpl client;
+	private final IShard shard;
+	private String name;
+	private final long id;
+	private final IGuild guild;
+	private int position;
+	private boolean isNSFW;
+
+	public final Cache<PermissionOverride> userOverrides;
+	public final Cache<PermissionOverride> roleOverrides;
+
+	public Category(IShard shard, String name, long id, IGuild guild, int position, boolean isNSFW, Cache<PermissionOverride> userOverrides, Cache<PermissionOverride> roleOverrides) {
+		this.shard = shard;
+		this.client = (DiscordClientImpl) shard.getClient();
+		this.name = name;
+		this.guild = guild;
+		this.position = position;
+		this.id = id;
+		this.isNSFW = isNSFW;
+		this.userOverrides = userOverrides;
+		this.roleOverrides = roleOverrides;
+	}
+
+	public void setName(String name) {
+		this.name = name;
+	}
+
+	public void setPosition(int position) {
+		this.position = position;
+	}
+
+	public void setNSFW(boolean isNSFW) {
+		this.isNSFW = isNSFW;
+	}
+
+	@Override
+	public void delete() {
+		PermissionUtils.requirePermissions(getModifiedPermissions(getClient().getOurUser()), EnumSet.of(Permissions.MANAGE_CHANNEL));
+
+		((DiscordClientImpl) client).REQUESTS.DELETE.makeRequest(DiscordEndpoints.CHANNELS+id);
+	}
+
+	@Override
+	public boolean isDeleted() {
+		return getClient().getCategoryByID(id) != this;
+	}
+
+	@Override
+	public IGuild getGuild() {
+		return guild;
+	}
+
+	@Override
+	public int getPosition() {
+		return position;
+	}
+
+	@Override
+	public void changePosition(int position) {
+		edit(new ChannelEditRequest.Builder().position(position).build());
+	}
+
+	private void edit(ChannelEditRequest request) {
+		PermissionUtils.requirePermissions(getModifiedPermissions(client.getOurUser()), EnumSet.of(Permissions.MANAGE_CHANNEL, Permissions.MANAGE_CHANNELS));
+
+		try {
+			client.REQUESTS.PATCH.makeRequest(
+					DiscordEndpoints.CHANNELS + id,
+					DiscordUtils.MAPPER.writeValueAsString(request));
+		} catch (JsonProcessingException e) {
+			Discord4J.LOGGER.error(LogMarkers.HANDLE, "Discord4J Internal Exception", e);
+		}
+	}
+
+	@Override
+	public String getName() {
+		return name;
+	}
+
+	@Override
+	public void changeName(String name) {
+		if (name == null || !name.matches("^[a-z0-9-_]{2,100}$"))
```

Comment:
Precompile and store the regex

Reply:
I'll do it if it's to also change the following to be be consistent with https://github.com/austinv11/Discord4J/blob/master/src/main/java/sx/blah/discord/handle/impl/obj/Channel.java#L636 and https://github.com/austinv11/Discord4J/blob/master/src/main/java/sx/blah/discord/handle/impl/obj/Channel.java#L644

Reply:
Sure